### DISCUSSION TITLE
Bloc - Clean Architeture

### DISCUSSION THREAD

**Discussion by marcelomoresco** [Flutterando/forum#34] [Perguntas e Respostas] [answered]
Boa Tarde pessoal, 

Queria saber aonde que fica o bloC seguindo a Clean Architeture? Na camada de presentator?

**Comment by mateusfccp** [comment]
Implementar Clean Architecture em Flutter é difícil. Isso porque, em Clean Architecture, propõe-se a inversão das dependências, de forma que o _interactor_ dependa do seu _output port_, implementado pelo _presenter_, e é o _interactor_ que chama esse _output port_.

![image](https://user-images.githubusercontent.com/4605213/191362252-a8fe6de2-b88b-4fc1-94eb-4e356e50b3d3.png)

Essa inversão é difícil de implementar no Flutter, não por questões técnicas, mas pela forma como usualmente trabalhamos com as _views_ e seus estados.

Em especial, o `flutter_bloc` estabelece uma estrutura específica que torna inviável implementar Clean Architecture. Isso porque normalmente emitimos estados de dentro do próprio `Bloc`/`Cubit`. Isso significa que o _interactor_ deixa de ter o controle de quando, ativamente, o _presenter_, através do _output port_, vai atualizar o estado da _view_.

O mais comum é usarmos o `Bloc`/`Cubit` (ou qual seja o "state manager") como tanto _presenter_ como _controller_, ao mesmo tempo. Não sei se isso é errado, por si só. No entanto, esta implementação normalmente é acompanhada da falta de inversão de dependência, já que este _controller_/_presenter_ que contém o _use case_. Veja, por exemplo, como isso acontece [nesse código]( https://github.com/Flutterando/clean-dart-search-bloc/blob/master/lib/app/search/presenter/search_bloc.dart).

Provavelmente exista uma forma de fazer isso, nunca trabalhei nessa ideia. No entanto, seria diferente de tudo o que vi até agora, já que todos que eu conheço usam o `flutter_bloc` conforme a documentação sugere, isto é, emitindo os valores de dentro do próprio `Bloc`/`Cubit`.

**Comment by eduardoflorence** [comment, marked as answer]
@marcelomoresco eu costumo sempre colocar na camada Presenter, pois ele representa o "estado" da View (page) que está nessa camada. Ele também é o responsável por interagir com a camada de domain, chamando os seus usecases.

**Comment by Bwolfs2** [comment]
É exatamente isso que o @eduardoflorence comentou, a camada View é composta tanto pela sua tela quando pelo Bloc que é o controlador dessa tela ou seja é apenas a primeira camada do Clean Arch
